cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/SparkBatchSinkFactory.java

```diff
@@ -82,22 +83,47 @@ private SparkBatchSinkFactory(Map<String, OutputFormatProvider> providers,
     this.sinkOutputs = sinkOutputs;
   }
 
-  void addOutput(String stageName, String outputName, OutputFormatProvider outputFormatProvider) {
-    if (outputFormatProviders.containsKey(outputName) || datasetInfos.containsKey(outputName)) {
-      throw new IllegalArgumentException("OutputFormatProvider with name " + outputName + " is already defined. " +
-                                           "Cannot add multiple OutputFormatProviders with the same name.");
+  void addOutput(String stageName, Output output) {
+    if (output instanceof Output.DatasetOutput) {
+      Output.DatasetOutput datasetOutput = (Output.DatasetOutput) output;
+      addOutput(stageName, datasetOutput.getName(), datasetOutput.getAlias(), datasetOutput.getArguments());
+    } else if (output instanceof Output.OutputFormatProviderOutput) {
+      Output.OutputFormatProviderOutput ofpOutput = (Output.OutputFormatProviderOutput) output;
+      addOutput(stageName, ofpOutput.getAlias(),
+                new BasicOutputFormatProvider(ofpOutput.getOutputFormatProvider().getOutputFormatClassName(),
+                                              ofpOutput.getOutputFormatProvider().getOutputFormatConfiguration()));
+    } else {
+      throw new IllegalArgumentException("Unknown output format type: " + output.getClass().getCanonicalName());
     }
-    outputFormatProviders.put(outputName, outputFormatProvider);
-    addStageOutput(stageName, outputName);
+  }
+
+  void addOutput(String stageName, String alias, OutputFormatProvider outputFormatProvider) {
+    addOutput(stageName, alias,
+              new BasicOutputFormatProvider(outputFormatProvider.getOutputFormatClassName(),
+                                            outputFormatProvider.getOutputFormatConfiguration()));
   }
 
   void addOutput(String stageName, String datasetName, Map<String, String> datasetArgs) {
-    if (outputFormatProviders.containsKey(datasetName) || datasetInfos.containsKey(datasetName)) {
-      throw new IllegalArgumentException("Output dataset with name " + datasetName + " is already defined. " +
+    addOutput(stageName, datasetName, datasetName, datasetArgs);
+  }
+
+  private void addOutput(String stageName, String alias,
+                         BasicOutputFormatProvider outputFormatProvider) {
+    if (outputFormatProviders.containsKey(alias) || datasetInfos.containsKey(alias)) {
+      throw new IllegalArgumentException("OutputFormatProvider with name " + alias + " is already defined. " +
```

Comment:
I think the message is very confusing, it exposes implementation, too. If the user calls addOutput(Output.ofDataset("x").alias("y") - assuming that it how it is supposed to be used? - but an output with alias "y" alsready exists, then the message is "OutputFormatProvider with name y is already defined. Cannot add multiple OutputFormatProviders with the same name". Even though what is clashing here is the alias, not the name. Also the message should just be in terms of "output" and perhaps it should mention the existing output that has the same alias.